README.md
# WP Local Docker

> WP Local Docker is an easy to use Docker based local development environment for WordPress development that works on Mac, Windows, and Linux.

[![Support Level](https://img.shields.io/badge/support-active-green.svg)](#support-level) [![Release Version](https://img.shields.io/github/release/10up/wp-local-docker-v2.svg)](https://github.com/10up/wp-local-docker-v2/releases/latest) [![MIT License](https://img.shields.io/github/license/10up/wp-local-docker-v2.svg)](https://github.com/10up/wp-local-docker-v2/blob/master/LICENSE)

## Documentation

### Introduction

WP Local Docker is an easy to use Docker based local development environment for WordPress development that works on Mac,
Windows, and Linux. Any number of environments can be created and may run at the same time<sup>[1](#faq)</sup>, and requests will be routed
appropriately to the correct environment based on the hostname specified during environment creation.

Each environment within WP Local Docker is powered by nginx, phpfpm, memcached, and if desired, elasticsearch. PHP versions
5.6, 7.0, 7.1, 7.2, 7.3, 7.4, or 8.0 are all supported. Supporting all environments within WP Local Docker is a MySQL container to run
all MySQL databases, WP Snapshots to easily push and pull snapshots of a WordPress installation, PHPMyAdmin to manage
MySQL databases with a familiar UI, and mailcatcher to catch any mail sent from all environments.

In addition to the services required to run WordPress, WP Local Docker will also download and install WordPress as a
single site installation, a Multisite with Subdirectories, a Multisite with Subdomains, or the core development version.

---

### Prerequisites

WP Local Docker requires [docker](https://www.docker.com/), [docker-compose](https://docs.docker.com/compose/), [Node](https://nodejs.org/en/), and npm. It is recommended that you use the latest versions of
docker and docker-compose. Node 12 is the minimum version of node that is currently supported. While WP Local Docker _may_ work with other versions of Node, compatibility is not guaranteed. We recommend using [NVM](https://github.com/nvm-sh/nvm) to manage Node versions.

#### MacOS

[Docker Desktop](https://docs.docker.com/docker-for-mac/install/) is available for download from the [Docker website](https://docs.docker.com/docker-for-mac/install/) and will
install docker-compose automatically. NodeJS and npm can be installed from the [NodeJS website](https://nodejs.org),
via a package manager, such as [Homebrew](https://brew.sh/), or using a version manager, such as
[nvm](https://github.com/creationix/nvm).

##### NodeJS EACCESS Error

If Node was installed via the download from the NodeJS website, you will likely get an `EACCESS` error when trying to install
global npm packages without using sudo. Npm has an article on [preventing permission errors](https://docs.npmjs.com/getting-started/fixing-npm-permissions#option-2-change-npms-default-directory-to-another-directory)
if you'd like to run the command without sudo. Alternatively, just run the install command with sudo.

#### Windows

[Docker Desktop](https://docs.docker.com/docker-for-windows/install/) is available for download from the [Docker website](https://docs.docker.com/docker-for-windows/install/) and will
install docker-compose automatically. NodeJS and npm can be installed from the [NodeJS website](https://nodejs.org). You may also need [Python](https://www.python.org/downloads/windows/) 3.7+ and [Visual Studio](https://visualstudio.microsoft.com/downloads/) 2015 or newer with the [“Desktop development with C++” workload](https://docs.microsoft.com/en-us/cpp/build/vscpp-step-0-installation?view=msvc-160).

It is recommended that you use the [WSL/2 backend for Docker](https://docs.docker.com/docker-for-windows/wsl/). You should use ([nvm](https://github.com/creationix/nvm) to install Node inside of your default Linux distro. Once you have, you can install WP Local Docker, from inside of Linux, following the [installation instructions](#Installation).

It is helpful to share git credentials between Windows and WSL/2. To do so, run the following, from inside of your default Linux, making sure to change `USER-NAME` to your Windows username:

```bash
git config --global credential.helper "/mnt/c/Program\ Files/Git/mingw64/libexec/git-core/git-credential-manager.exe"
cd ~/.ssh
cp /mnt/c/Users/USER-NAME/.ssh/id_rsa* .
```
##### HTTPS on Windows
If https is not working in the browsers on Windows you will need to setup a shared root certificate.
You will need [mkcert installed](https://github.com/FiloSottile/mkcert) inside WSL2 machine. With `mkcert` installed run:

```bash
mkcert -CAROOT
```

That will output the path to a certificate root directory which contains the rootCA.pem file you will need to share with Windows.
List the directory path you got from the previous command:

```bash
ls /home/<user>/.local/share/mkcert
```

If there is no such directory or the directory is empty run:

```bash
mkcert -install
```

You will get output like:

```bash
Created a new local CA 💥
The local CA is now installed in the system trust store! ⚡️
The local CA is now installed in the Firefox trust store (requires browser restart)! 🦊
```

If there was no root certificate initially, you will need to regenerate the certificates for all the sites:

```bash
10updocker cert generate <yourwebsite.test>
```

Now there should be a rootCA.pem file in the root certicicate directory. Copy the rootCA.pem into any directory on Windows. In the Windows PowerShell set that directory as `CAROOT` environment variable:

```bash
$Env:CAROOT = "C:\path\to\directory\with\certificate"
```

Then install certificate in the Windows PowerShell with:

```bash
mkcert -install
```

You may need to install `mkcert` on Windows, or just use the latest executable from [the releases page](https://github.com/FiloSottile/mkcert/releases).

You will need to restart any open browsers for this change to take effect!
That should fix the https issues.

#### Linux

Docker has platform specific installation instructions available for linux on their [documentation site](https://docs.docker.com/install/#supported-platforms).
Once docker is installed, you will need to [manually install docker compose](https://docs.docker.com/compose/install/).
NodeJS can be installed via a package manager for many linux platforms [following these instructions](https://nodejs.org/en/download/package-manager/).
Setup c compilers & add sudo user to docker group:

```
sudo apt install -y gcc make
sudo usermod -a -G docker user
```

---

## Installation

Once all installation prerequisites have been met, WP Local Docker is installed as a global npm package by running
`npm install -g wp-local-docker`. You can confirm it has been installed by running `10updocker --version`.

### Configuration

The first time you run a WP Local Docker command, default configuration settings will be used if you have not manually
configured WP Local Docker beforehand. By default, WP Local Docker will store all environments within the
`~/wp-local-docker-sites` directory and try to manage your hosts file when creating and deleting environments. If you
would like to customize the environment path or opt to not have WP Local Docker update your hosts file, run
`10updocker configure` and follow the prompts.

### Updating

To update WP Local Docker, run `npm install -g wp-local-docker` again and NPM will install the latest version over your current one.

---

## Using WP Local Docker

### Create an Environment

`10updocker create` will present you with a series of prompts to configure your environment to suit your needs.

It is recommended that you use the `.test` top level domain (TLD) for your local environments, as this TLD is reserved
for the testing of software and is not intended to ever be installed into the global Domain Name System. Additionally,
WP Local Docker is configured to send any container to container traffic for .test TLDs directly to the gateway
container, so that things like WP Cron and the REST API can work between environments out of the box.

### Migrate a WP Local Docker V1 Environment

`10updocker migrate <OLD_PATH> [NEW_ENV]` will migrate an old standalone WP Local Docker environment into a new WP Local Docker V2
environment. Before running this command, create a new environment using the `10updocker create` command.

`OLD_PATH` should be the path to the root of your old WP Local Docker environment.

`NEW_ENV` should specify what environment to import into. If omitted, you will be prompted to select from available environments

Example:

* `10updocker migrate ~/sites/mysite`

### Migrate a WP Local Docker V2 Environment to v3

`10updocker configure` to update to the latest `.wplocaldocker/global/docker-compose.yml` and answer `yes` to `Do you want to reset your global services configuration? This will reset any customizations you have made.`

### M1 Architecture consideration

To utilize the new M1 architecture optimised docker images on existing sites, edit the `docker-compose.yml` file to match a newly created on. Alternatively you can just create new sites and they will be optimised for M1 archtecture.

### Delete an Environment

`10updocker delete <hostname>` will delete an environment with the given hostname. Any local files, docker volumes, and
databases related to the environment will be deleted permanently.

A special hostname `all` is available that will delete all environments. You will be asked to confirm deletion of each
environment.

### Stop an Environment

`10updocker stop <hostname>` will stop an environment from running while retaining all files, docker volumes, and
databases related to the environment.

A special hostname `all` is available that will stop all running environments as well as the global services.

### Start an Environment

`10updocker start <hostname>` will start a preexisting environment.

A special hostname `all` is available that will start all environments as well as the global services.

### Configure an Environment

An environment's `config/` directory contains settings for various services:

- `elasticsearch/`
    - `elasticsearch.yml`
- `nginx/`
    - `default.conf`
    - `develop.conf`
    - `server.conf`
- `php-fpm/`
    - `docker-php-ext-xdebug.ini` — this is equivalent to `php.ini`, and is where you should put settings that would be found in a file named `php.ini`, such as `max_execution_time`, `memory_limit`, `xdebug.max_nesting_level`, `post_max_size`, `upload_max_filesize`, and similar.
    - `wp-cli.develop.yml`
    - `wp-cli.local.yml`

After making changes to files in `config/` or to `docker-compose.yml`, you must recreate the containers. Instructions for doing so can be found on this page under [Restart an environment](#restart-an-environment)

### Restart an Environment

`10updocker restart <hostname>` will restart all services associated with a preexisting environment.

A special hostname `all` is available that will restart all environments as well as the global services.

Restarting with `10updocker restart` does not apply configuration changes. If you've made changes to an environment's `docker-compose.yml` or the files in `/config` directory, you will need to stop the environment (but not global services) and then run [`docker-compose up`](https://docs.docker.com/compose/reference/up/) to recreate the containers:

0. Navigate to the root directory of the install you want to make changes to.
1. Make your changes to the configuration files.
2. Run `10updocker stop all` to stop all environments, including the global services.
3. Run `10updocker start` to get this install and the global services running.
4. Run `10updocker stop` to stop this install but not the global services.
5. Run [`docker-compose up`](https://docs.docker.com/compose/reference/up/) to recreate and start this install's docker containers. Wait for it to finish starting; this can be determined by watching the console output. Don't load the site in a browser. If there hasn't been anything new on the console for at least 10 seconds, you can proceed to the next step.
6. Kill the newly-started docker process with the `Control-C` keyboard command, or your system's equivalent instruction.
7. Run `10updocker restart` to restart the container.

### Upgrade an Environment

`10updocker upgrade <hostname>` will upgrade all services associated with a preexisting environment.

This command will assist you with keeping your environments up to date with the most recent upstream changes.

If the environment was created before `v2.6.0`, we recommend upgrading your environment for a noticeable
performance increase.

If the environment was created before `v3.0.1`, this will update the elasticsearch image. When updating the elasticsearch image, we need to delete the docker volume so you will need to reindex after running this command.

### Managing Certificates

`10updocker cert generate <hostname>` will generate certificates for use of SSL on local environments.

`10updocker cert install` installs a new certificate authority in the system trust store.

### Elasticsearch

If you have enabled Elasticsearch for a particular environment, you can send requests from the host machine to the
Elasticsearch server by prefixing the url path with `/__elasticsearch/`. For example, if you wanted to hit the
`/_all/_search/` endpoint of Elasticsearch, the URL would look like: `http://<hostname>/__elasticsearch/_all/_search`

In the ElasticPress settings, you should enter `http://<hostname>/__elasticsearch` as the Elasticsearch host.

### WP Snapshots

See the section on [using Snapshots](#user-content-snapshots)

### Running WP CLI Commands

Running WP CLI commands against an environment is easy. First, make sure you are somewhere within your environment
directory (by default, this is somewhere within `~/wp-local-docker-sites/<environment>/`). Once within the environment
directory, simply run `10updocker wp <command>`. `<command>` can be any valid command you would otherwise pass directly
to WP CLI.

Examples:


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file, per [the Keep a Changelog standard](http://keepachangelog.com/).

## [4.0.0] - 2023-07-06
### Added
- Adds the option to select PHP 8.1 or 8.2 when creating a new environment.
- Adds support for proxying media from a HTTPS upstream utilising SNI.

### Changed
- Updated the snapshot system to utilise the new PHP agnostic [Snapshots](https://github.com/10up/snapshots) WP-CLI package.
- Updated the `10updocker restart` command to actually remove and rebuild the environments, ensuring that config changes are applied.

### Fixed
- Fixes an issue where a space in the mkcert path could cause certs to fail to generate and hosts to fail to update.
- Fixes the external network deprecation notice for new environments.

### Removes
- Removes `10updocker init` and `10updocker clone` commands

## [3.0.2] - 2022-02-02
### Added
- Adds warning for EOL versions of PHP.

### Fixed
- Fixed log file permissions on linux platforms

## [3.0.1] - 2022-01-05

### Fixed
- Fixes the correct image name for phpfpm in Linux platforms
- Fixes issue with php ini file being in the wrong location and add upgrade command to resolve this for exisitng projects.
- Fixes xdebug values for xdebug3 since the port and settings changed moving from v2 to v3.

### Changed
- Update elasticsearch image to 7.9.3
- Add Elasticsearch image update to the `upgrade` command. This will also delete the docker volume since it can cause issues. (Prompts user before deleting data.)
- More readme updates for upgrading to the latest images.

## [3.0.0] - 2021-11-11
### Added
- Adds support for PHP 8.0.
- Adds support for ARM images.

### Changed
- Updates dependencies to their latest versions.
- Suppresses WP-CLI errors during removing the default content.

### Fixed
- Fixes issues with some docker-compose commands caused by `docker-compose` library.
- Fixes phpfpm image name issue that occurs on Linux if the user has a non-standard name.
- Fixes empty content issues on create.
- Fixes issue with missing npm request package.

### Removes
- Removes `10updocker init` and `10updocker clone` commands

## [2.8.1] - 2021-01-29
### Changed
- Compose files have been updated to build PHP images in the `.containers` context.

### Fixed
- Fixed unexpected behaviour caused by errors thrown from the `execSync` command when a subcommand exited with non-zero exit code.

## [2.8.0] - 2021-01-26
### Added
- Adds `clone` command to clone git repositories and create new environments for it.
- Adds `completion <shell>` command to display instructions how to activate autocomplete for a specific shell. Supported shells : `bash` and `zsh`.
- Adds the ability to use project specific configuration files. It is read during clonning process and allows to define a configuration required for the project and allows to omit setup questions. If the configuration is not sufficient (for example, it misses php version), then appropriate questions will be asked fill gaps.
- Adds `init` command to generate project configuration files.
- Adds the ability to update a docker-compose config for an environment if the configuration file contains `dockerCompose` callback function.
- Adds global `--env=<environment>` option to specify an environment to use and to skip environment selection prompt.
- Adds global `--verbose` option to switch between quite and verbose output modes.
- Adds `--tail=<number>` argument to the `logs` command to limit the number of lines to show from the end of the logs for each container. It works the same way as it works for `docker-compose logs` itself and has `all` as default.
- Adds `--pull` option to the `start` command to automatically pull latest images before an environment starts.
- Adds `--remove-built-images` flag to the `image update` command to force removal of built images.
- Adds `--yes` flag to the `image update` command to suppress questions.
- Adds `<cmd>` positional argument to the `shell` command to override command to run in the container. By default it is still `bash` but now it is possible to run different single commands without launching bash first.
- Adds `list` and `ls` commands to list all the environments and meta information.
- Adds engines information to the package.json to strictly denote that Node v12+ is required.
- Adds `cert install` command to install a new local CA in the system trust store.
- Adds `cert generate <domains..>` command to generate self-signed SSL certificates for certain domains.
- Auto-completion script for `10updocker` and `10updocker-hosts` commands (props [eugene-manuilov](https://github.com/eugene-manuilov) via [#105](https://github.com/10up/wp-local-docker-v2/pull/105))
- Adds `SYS_PTRACE` capability to the php service to support [`strace`](https://linux.die.net/man/1/strace) command.

### Changed
- Reworks all commands to use `yargs` cli framework.
- Updates the way how `wp` and `wpsnapshots` commands determine subcommand arguments that needs to be passed into containers. Previously, wp-local-docker passed all arguments to the underlying command even if some arguments weren't intended for it. Now it passes everything that comes after the command keyword only, for example: `10updocker --env=my-site-test wp plugin install hello-dolly --version=1.7.2` runs `wp plugin install hello-dolly --version=1.7.2` command in the container of the `my-site-test` environment and doesn't pass `--env=my-site-test` argument itself.
- Removes https question and updates `create` command to create sites with HTTPS only.
- Updates bash completion script to support new commands and aliases.
- `ini` comments to use semi-colon for compatibility (props [@aaemnnosttv](https://github.com/aaemnnosttv) via [#113](https://github.com/10up/wp-local-docker-v2/pull/113))
- Documentation updates (props [@jeffpaul](https://github.com/jeffpaul) via [#58](https://github.com/10up/wp-local-docker-v2/pull/58))

src/create.js
const commandUtils = require( './command-utils' );
const path = require( 'path' );
const fs = require( 'fs-extra' );
const yaml = require( 'write-yaml' );
const inquirer = require( 'inquirer' );
const promptValidators = require( './prompt-validators' );
const database = require( './database' );
const gateway = require( './gateway' );
const environment = require( './environment.js' );
const wordpress = require( './wordpress' );
const envUtils = require( './env-utils' );
const sudo = require( 'sudo-prompt' );
const config = require( './configure' );
const chalk = require( 'chalk' );
const os = require( 'os' );
const { images } = require( './image' );

const help = function() {
    const help = `
Usage: 10updocker create

Creates a new docker environment interactively.
`;
    console.log( help );
    process.exit();
};

const createEnv = async function() {
    const baseConfig = {
        // use version 2 so we can use limits
        'version': '2.2',
        'services': {
            'memcached': {
                'image': images['memcached'],
            },
            'nginx': {
                'image': images['nginx'],
                'expose': [
                    '80',
                    '443'
                ],
                'volumes': [
                    './wordpress:/var/www/html:cached'
                ],
                'depends_on': [
                    'phpfpm'
                ],
                'networks': [
                    'default',
                    'wplocaldocker'
                ],
                'environment': {
                    'CERT_NAME': 'localhost',
                    'HTTPS_METHOD': 'noredirect'
                }
            }
        }
    };

    const networkConfig = {
        'networks': {
            'wplocaldocker': {
                'external': {
                    'name': 'wplocaldocker'
                }
            }
        }
    };

    const volumeConfig = {
        'volumes': {}
    };
    volumeConfig.volumes[ envUtils.cacheVolume ] = {
        'external': {
            'name': `${envUtils.cacheVolume}`
        }
    };

    const questions = [
        {
            name: 'hostname',
            type: 'input',
            message: 'What is the primary hostname for your site? (Ex: docker.test)',
            validate: promptValidators.validateNotEmpty,
            filter: promptValidators.parseHostname,
        },
        {
            name: 'addMoreHosts',
            type: 'confirm',
            message: 'Are there additional domains the site should respond to?',
            default: false,
        },
        {
            name: 'extraHosts',
            type: 'input',
            message: 'Enter additional hostnames separated by spaces (Ex: docker1.test docker2.test)',
            filter: async function( value ) {
                const answers = value.split( ' ' ).map( function( value ) {
                    return value.trim();
                } ).filter( function( value ) {
                    return value.length > 0;
                } ).map( promptValidators.parseHostname );

                return answers;
            },
            when: function( answers ) {
                return answers.addMoreHosts === true;
            }
        },
        {
            name: 'addHttps',
            type: 'confirm',
            message: 'Do you want to enable HTTPS?',
            default: false,
        },
        {
            name: 'mediaProxy',
            type: 'confirm',
            message: 'Do you want to set a proxy for media assets? (i.e. Serving /uploads/ directory assets from a production site)',
            default: false,
        },
        {
            name: 'proxy',
            type: 'input',
            message: 'Proxy URL',
            default: function( answers ) {
                return envUtils.createDefaultProxy( answers.hostname );
            },
            validate: promptValidators.validateNotEmpty,
            filter: promptValidators.parseProxyUrl,
            when: function( answers ) {
                return answers.mediaProxy === true;
            }
        },
        {
            name: 'phpVersion',
            type: 'list',
            message: 'What version of PHP would you like to use?',
            choices: [ '7.4', '7.3', '7.2', '7.1', '7.0', '5.6' ],
            default: '7.3',
        },
        {
            name: 'elasticsearch',
            type: 'confirm',
            message: 'Do you need Elasticsearch',
        },
        {
            name: 'wordpress',
            type: 'confirm',
            message: 'Do you want to install WordPress?',
        },
        {
            name: 'wordpressType',
            type: 'list',
            message: 'Select a WordPress installation type:',
            choices: [
                { name: 'Single Site', value: 'single' },
                { name: 'Subdirectory Multisite', value: 'subdirectory' },
                { name: 'Subdomain Multisite', value: 'subdomain' },
                { name: 'Core Development Version', value: 'dev' },
            ],
            default: 'single',
            when: function( answers ) {
                return answers.wordpress === true;
            }
        },
        {
            name: 'emptyContent',
            type: 'confirm',
            message: 'Do you want to remove the default content?',
            when: function( answers ) {
                return answers.wordpress === true;
            }
        },
        {
            name: 'title',
            type: 'input',
            message: 'Site Name',
            default: function( answers ) {
                return answers.hostname;
            },
            validate: promptValidators.validateNotEmpty,
            when: function( answers ) {
                return answers.wordpress === true;
            }
        },
        {
            name: 'username',
            type: 'input',
            message: 'Admin Username',
            default: 'admin',
            validate: promptValidators.validateNotEmpty,
            when: function( answers ) {
                return answers.wordpress === true;
            }
        },
        {
            name: 'password',
            type: 'input',
            message: 'Admin Password',
            default: 'password',
            validate: promptValidators.validateNotEmpty,
            when: function( answers ) {
                return answers.wordpress === true;
            }
        },
        {
            name: 'email',
            type: 'input',
            message: 'Admin Email',
            default: 'admin@example.com',
            validate: promptValidators.validateNotEmpty,
            when: function( answers ) {
                return answers.wordpress === true;
            }
        },
    ];

    const answers = await inquirer.prompt( questions );

    // Folder name inside of /sites/ for this site
    const envHost = answers.hostname;
    const envSlug = envUtils.envSlug( envHost );
    const envPath = await envUtils.envPath( envHost );

    // Default nginx configuration file
    let nginxConfig = 'default.conf';

    if ( await fs.exists( envPath ) === true ) {
        console.log();
        console.error( `Error: ${envHost} environment already exists. To recreate the environment, please delete it first by running \`10updocker delete ${envHost}\`` );
        process.exit( 1 );
    }

    await gateway.startGlobal();

    let allHosts = [ answers.hostname ];
    const starHosts = [];

    if ( answers.addMoreHosts === true ) {
        answers.extraHosts.forEach( function( host ) {
            allHosts.push( host );
        } );
    }

    // Remove duplicates
    allHosts = allHosts.filter( function( item, pos, self ) {
        return self.indexOf( item ) === pos;
    } );

    allHosts.forEach( function( host ) {
        starHosts.push( `*.${host}` );
    } );

    // Additional nginx config based on selections above
    baseConfig.services.nginx.environment.VIRTUAL_HOST = allHosts.concat( starHosts ).join( ',' );

    baseConfig.services.phpfpm = {
        'image': images[`php${answers.phpVersion}`],
        'environment': {
            'ENABLE_XDEBUG': 'false'
        },
        'volumes': [
            './wordpress:/var/www/html:cached',
            './config/php-fpm/docker-php-ext-xdebug.ini:/etc/php.d/docker-php-ext-xdebug.ini:cached',
            `${envUtils.cacheVolume}:/var/www/.wp-cli/cache:cached`,
        ],
        'depends_on': [
            'memcached',
        ],
        'networks': [
            'default',
            'wplocaldocker'
        ],
        'dns': [
            '10.0.0.2'
        ]
    };

    // Unlike Mac and Windows, Docker is a first class citizen on Linux
    // and doesn't have any kind of translation layer between users and the
    // file system. Because of this the phpfpm container will be running as the 
    // wrong user. Here we setup the docker-compose.yml file to rebuild the
    // phpfpm container so that it runs as the user who created the project.
    if ( os.platform() == 'linux' ) {
        baseConfig.services.phpfpm.image = `wp-php-fpm-dev-${answers.phpVersion}-${process.env.USER}`;
        baseConfig.services.phpfpm.build = {
            'dockerfile': '.containers/php-fpm',
            'context': '.',
            'args': {
                'PHP_IMAGE': images[`php${answers.phpVersion}`],
                'CALLING_USER': process.env.USER,
                'CALLING_UID': process.getuid()
            }
        };
        baseConfig.services.phpfpm.volumes.push( `~/.ssh:/home/${process.env.USER}/.ssh:cached` );
    }
    else {
        // the official containers for this project will have a www-data user. 
        baseConfig.services.phpfpm.volumes.push( '~/.ssh:/home/www-data/.ssh:cached' );
    }

    if ( answers.wordpressType == 'dev' ) {
        baseConfig.services.phpfpm.volumes.push( './config/php-fpm/wp-cli.develop.yml:/var/www/.wp-cli/config.yml:cached' );
        nginxConfig = 'develop.conf';
    } else {
        baseConfig.services.phpfpm.volumes.push( './config/php-fpm/wp-cli.local.yml:/var/www/.wp-cli/config.yml:cached' );
    }

    // Map the nginx configuraiton file
    baseConfig.services.nginx.volumes.push( `./config/nginx/${  nginxConfig  }:/etc/nginx/conf.d/default.conf:cached` );

    if ( answers.elasticsearch === true ) {
        baseConfig.services.phpfpm.depends_on.push( 'elasticsearch' );

        baseConfig.services.elasticsearch = {
            image: images['elasticsearch'],
            'expose': [
                '9200'
            ],
            'volumes': [
                './config/elasticsearch/elasticsearch.yml:/usr/share/elasticsearch/config/elasticsearch.yml:cached',
                './config/elasticsearch/plugins:/usr/share/elasticsearch/plugins:cached',
                'elasticsearchData:/usr/share/elasticsearch/data:delegated'
            ],
            'mem_limit': '1280M',
            'mem_reservation': '1280M',
            'environment': {
                ES_JAVA_OPTS: '-Xms512m -Xmx512m'
            }
        };

        volumeConfig.volumes.elasticsearchData = {};
    }


    // Create webroot/config
    console.log( 'Copying required files...' );

    await fs.ensureDir( path.join( envPath, 'wordpress' ) );
    await fs.copy( path.join( envUtils.srcPath, 'config' ), path.join( envPath, 'config' ) );
    await fs.ensureDir( path.join( envPath, '.containers' ) );
    await fs.copy( path.join( envUtils.srcPath, 'containers' ), path.join( envPath, '.containers' ) );

    // Write Docker Compose
    console.log( 'Generating docker-compose.yml file...' );
    const dockerCompose = Object.assign( baseConfig, networkConfig, volumeConfig );
    await new Promise( resolve => {
        yaml( path.join( envPath, 'docker-compose.yml' ), dockerCompose, { 'lineWidth': 500 }, function( err ) {
            if ( err ) {
                console.log( err );
            }
            console.log( 'done' );
            resolve();
        } );
    } );

    // Write wp-cli config
    console.log( 'Generating wp-cli.yml file...' );
    await new Promise( resolve => {
        yaml(
            path.join( envPath, 'wp-cli.yml' ),
            {
                ssh: 'docker-compose:phpfpm'
            },
            {
                lineWidth: 500
            },
            function( err ) {
                if ( err ) {
                    console.error( err );
                }
                console.log( 'done' );
                resolve();
            }
        );
    } );

    // Media proxy is selected
    if ( answers.mediaProxy === true ) {
        // Write the proxy to the config files
        console.log( 'Writing proxy configuration...' );

        await new Promise( resolve => {
            fs.readFile( path.join( envPath, 'config', 'nginx', nginxConfig ), 'utf8', function( err, curConfig ) {
                if ( err ) {
                    console.error( `${ chalk.bold.yellow( 'Warning: ' ) }Failed to read nginx configuration file. Your media proxy has not been set. Error: ${ err }` );
                    resolve();
                    return;
                }

                fs.writeFile( path.join( envPath, 'config', 'nginx', nginxConfig ), config.createProxyConfig( answers.proxy, curConfig ), 'utf8', function ( err ) {
                    if ( err ) {
                        console.error( `${ chalk.bold.yellow( 'Warning: ' ) }Failed to write configuration file. Your media proxy has not been set. Error: ${ err }` );
                        resolve();
                        return;
                    }
                } );

                console.log( 'Proxy configured' );
                resolve();
            } );
        } );
    }

    // Create database
    console.log( 'Creating database' );
    await database.create( envSlug );
    await database.assignPrivs( envSlug );

    await environment.start( envSlug );

    if ( answers.wordpress === true ) {
        if ( answers.wordpressType === 'dev' ) {
            await wordpress.downloadDevelop( envSlug );
        } else {
            await wordpress.download( envSlug );
        }

        await wordpress.configure( envSlug );

        await wordpress.install( envSlug, envHost, answers );

        await wordpress.setRewrites( envSlug );

        if ( answers.emptyContent === true ) {
            await wordpress.emptyContent( envSlug );
        }
    }

    if ( await config.get( 'manageHosts' ) === true ) {
        console.log( 'Adding entry to hosts file' );
        const sudoOptions = {
            name: 'WP Local Docker'
        };
        await new Promise( resolve => {
            const hostsstring = allHosts.join( ' ' );
            sudo.exec( `10updocker-hosts add ${hostsstring}`, sudoOptions, function( error, stdout ) {
                if ( error ) {
                    console.error( `${ chalk.bold.yellow( 'Warning: ' ) }Something went wrong adding host file entries. You may need to add the /etc/hosts entries manually.` );
                    resolve();
                    return;
                }

                console.log( stdout );
                resolve();
            } );
        } );
    }

    // Track things we might need to know later in order to clean up the environment
    const envConfig = {
        'envHosts': allHosts
    };
    await fs.writeJson( path.join( envPath, '.config.json' ), envConfig );

    console.log();
    console.log( 'Successfully Created Site!' );

    if ( answers.wordpressType === 'subdomain' ) {
        console.log( 'Note: Subdomain multisites require any additional subdomains to be added manually to your hosts file!' );
    }
};

const command = async function() {
    switch ( commandUtils.subcommand() ) {
        case 'help':
            help();
            break;
        default:
            await createEnv();
            break;
    }
};

module.exports = { command, help };
